src/ChainSafe.Gaming.MetaMask.Unity/MetaMaskTransactionExecutor.cs

```diff
@@ -0,0 +1,94 @@
+
+        private readonly ISigner signer;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MetaMaskTransactionExecutor"/> class.
+        /// </summary>
+        /// <param name="logWriter">Log Writer used for logging messages and errors.</param>
+        /// <param name="metaMaskProvider">Metamask provider that connects to Metamask and makes JsonRPC requests.</param>
+        /// <param name="signer">Signer for fetching address.</param>
+        public MetaMaskTransactionExecutor(ILogWriter logWriter, IMetaMaskProvider metaMaskProvider, ISigner signer)
+        {
+            this.logWriter = logWriter;
+
+            this.metaMaskProvider = metaMaskProvider;
+
+            this.signer = signer;
+        }
+
+        /// <summary>
+        /// Implementation of <see cref="ITransactionExecutor.SendTransaction"/>.
+        /// Send a transaction using Metamask.
+        /// This prompts user to approve a transaction on Metamask.
+        /// </summary>
+        /// <param name="transaction">Transaction to send.</param>
+        /// <returns>Hash response of a successfully executed transaction.</returns>
+        /// <exception cref="Web3Exception">Throws Exception if executing transaction fails.</exception>
+        public async Task<TransactionResponse> SendTransaction(TransactionRequest transaction)
+        {
+            if (string.IsNullOrEmpty(transaction.From))
+            {
+                transaction.From = await signer.GetAddress();
+            }
+
+            logWriter.Log($"fee stuff {transaction.MaxFeePerGas} {transaction.MaxPriorityFeePerGas}");
+
+            TransactionInput transactionInput = new TransactionInput
+            {
+                From = transaction.From,
+                To = transaction.To,
+                Gas = transaction.GasLimit,
+                GasPrice = transaction.GasPrice,
+                Value = transaction.Value,
+                Data = transaction.Data ?? "0x",
+                Nonce = transaction.Nonce,
+                AccessList = transaction.AccessList,
+            };
+
+            string hash = await metaMaskProvider.Request<string>("eth_sendTransaction", transactionInput);
+
+            // TODO replace validation with regex
+            if (!hash.StartsWith("0x") || hash.Length != 66)
+            {
+                throw new Web3Exception($"incorrect txn response format {hash}");
+            }
+
+            logWriter.Log($"Transaction executed with hash {hash}");
+
+            return await metaMaskProvider.Request<TransactionResponse>("eth_getTransactionByHash", hash);
+        }
+
+        /// <summary>
+        /// Implementation of <see cref="ILifecycleParticipant.WillStartAsync"/>.
+        /// Lifetime event method, called during initialization.
+        /// </summary>
+        /// <returns>async awaitable task.</returns>
+        public ValueTask WillStartAsync() => new ValueTask(Task.CompletedTask);
```

Comment:
You can remove ILifecycleParticipant from the list of parents, as there is no actual function to it.

Reply:
Also, in other components too.

Reply:
yeah you're right, I thought about that too but don't you think it could come in handy in the future?

Reply:
We can always add it when we need it.

The rule of thumb here is to keep only the stuff that is currently being used. Again we can add what will come in handy in the future, in the future. But this way we'll keep the project clean from the stuff that never came in handy eventually.

Reply:
alright, will do